Add 7.15.0 blog post (#2557)
babel/website: website/blog/2021-07-26-7.15.0.md

If you or your company want to support Babel and the evolution of JavaScript, but aren't sure how, you can donate to us on our [Open Collective](https://opencollective.com/babel) and, better yet, work with us on the implementation of [new ECMAScript proposals](https://github.com/babel/proposals) directly! As a volunteer-driven project, we rely on the community's support to fund our efforts in supporting the wide range of JavaScript users. Reach out at [team@babeljs.io](mailto:team@babeljs.io) if you'd like to discuss more!

## Highlights

### ECMAScript features enabled by default

In the last two meetings, the [top-level `await`](https://github.com/tc39/proposal-top-level-await) and [ergonomic brand checks for private fields](https://github.com/tc39/proposal-private-fields-in-in) proposals reached Stage 4.

```javascript
import * as db from "database";

await db.connect(); // top-level await

class DBConnector {
  #password;
  static isConnector(obj) {
    return #password in obj; // ergonomic brand checks
  }
}
```

Babel now supports them by default, so you can remove the following plugins from your configuration:

> - `@babel/plugin-syntax-top-level-await`

Reviewer: cool, can move these in website https://github.com/babel/website/blob/d12f0a40b4bf839ee673fce964d2088743d64d82/website/sidebars.json#L66-L69 and add note to the page like in

> **NOTE**: This plugin is included in `@babel/preset-env`

Author: I pushed it to https://github.com/babel/website/pull/2556, since I was doing some related changes.

If you want a more optimized output similar to the default code produced by TypeScript, you can enable the [`optimizeConstEnums`](TODO: Link docs) option of `@babel/plugin-tranform-typescript` or `@babel/preset-typescript`.

```typescript
// Input
const enum Animals { Dog }
console.log(Animals.Dog);

// Output (default)
var Animals;
(function (Animals) {
  Animals[Animals["Dog"] = 0] = "Dog";
})(Animals || (Animals = {}));

console.log(Animals.Dog);

// Output with `optimizeConstEnums`
console.log(0);
```
### Hack-style pipeline operator support ([#13191](https://github.com/babel/babel/pull/13191), [#13416](https://github.com/babel/babel/pull/13416))

["Hack-style pipelines"](https://github.com/js-choi/proposal-hack-pipes/) is a new flavor of the [pipeline operator](https://github.com/tc39/proposal-pipeline-operator) proposal, intended to replace the "smart mix" variant.

Hack-style pipelines require you to _always_ use a "topic token" (such as `#`), to reference the value of the previous pipeline step:

```javascript

> "World"
>   |> `Hello, ${#}!`
>   |> alert(#);

Reviewer 2: ```suggestion
// Input
"World"
  |> `Hello, ${#}!`
  |> alert(#);

// output
var _ref, _ref2;

_ref2 = (_ref = "World", `Hello, ${_ref}!`), alert(_ref2);
```

Added output so people can have glimpse about what the pipeline proposal does.

---
layout: post
title:  "7.15.0 Released: Hack-style pipelines, TypeScript const enums and Rhino target support"
author: Babel Team
authorURL: https://twitter.com/babeljs
date:   2021-07-26 0:00:00
categories: announcements
share_text: "Babel 7.15.0 Released"
---

This release enables parsing top-level await (Stage 4 at the [May](https://github.com/tc39/agendas/blob/master/2021/05.md) meeting) and transforming ergonomic brand checks for private fields (Stage 4 at the [July](https://github.com/tc39/agendas/blob/master/2021/07.md) meeting) by default. There is also now support for the [Hack-style](https://github.com/js-choi/proposal-hack-pipes) pipeline operator. We also improved our TypeScript support, implementing transform support for `const` enums and namespace aliases, and we expanded our heuristics to add `.displayName` to React components created by `React.createContext()` ([#13501](https://github.com/babel/babel/pull/13501)).

> We also introduced a new compiler assumption, [`noIncompleteNsImportDetection`](TODO: Link), to produce a smaller output when compiling ECMAScript modules to CommonJS without worrying about partially initialized namespace imports caused by module cycles.

Author: ```suggestion
We also introduced a new compiler assumption, [`noIncompleteNsImportDetection`](https://babeljs.io/docs/en/assumptions#noincompletensimportdetection), to produce a smaller output when compiling ECMAScript modules to CommonJS without worrying about partially initialized namespace imports caused by module cycles.
```

Additionally, you can now specify [Rhino](https://mozilla.github.io/rhino/) as a compilation target.

You can read the whole changelog [on GitHub](https://github.com/babel/babel/releases/tag/v7.15.0).

<!-- truncate -->

> We're really grateful for all the support the community has shown over the last month's since our [funding post](https://babeljs.io/blog/2021/05/10/funding-update) update in May. Reach out at [team@babeljs.io](mailto:team@babeljs.io) if you'd like to discuss about sponsorships!

Author: ```suggestion
We're really grateful for all the support the community has shown over the last months, since our [funding post](https://babeljs.io/blog/2021/05/10/funding-update) update in May. Reach out at [team@babeljs.io](mailto:team@babeljs.io) if you'd like to discuss about sponsorships!
```

- `@babel/plugin-proposal-private-property-in-object`

Please note that Babel can currently only parse top-level `await` and cannot transform it.

### New TypeScript features ([#13324](https://github.com/babel/babel/pull/13324), [#13528](https://github.com/babel/babel/pull/13528))

[TypeScript 4.4](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/) doesn't include any new syntax that we had to implement, other than a minor restriction: you cannot specify the value of an abstract class field.

```typescript=
abstract class C {
    abstract prop = 1; // now a SyntaxError!
}
```

However, we did add two TypeScript features that we have been missing for a long time: `const` enums and namespace aliases (`import Alias = Namespace`).

Previously we presented an error when using `const` enums since it requires type information to compile correctly. As a workaround, the community built plugins such as [`babel-plugin-const-enum`](https://github.com/dosentmatter/babel-plugin-const-enum#readme). Babel now ignores the `const` modifier when compiling enums, which matches TypeScript's behavior when using the `--isolatedModules` option.

> If you want a more optimized output similar to the default code produced by TypeScript, you can enable the [`optimizeConstEnums`](TODO: Link docs) option of `@babel/plugin-tranform-typescript` or `@babel/preset-typescript`.

Author: ```suggestion
If you want a more optimized output similar to the default code produced by TypeScript, you can enable the [`optimizeConstEnums`](https://babeljs.io/docs/en/babel-preset-typescript#optimizeconstenums) option of `@babel/plugin-tranform-typescript` or `@babel/preset-typescript`.
```

```typescript
// Input
const enum Animals { Dog }
console.log(Animals.Dog);

// Output (default)
var Animals;
(function (Animals) {
  Animals[Animals["Dog"] = 0] = "Dog";
})(Animals || (Animals = {}));

console.log(Animals.Dog);

// Output with `optimizeConstEnums`
console.log(0);
```
### Hack-style pipeline operator support ([#13191](https://github.com/babel/babel/pull/13191), [#13416](https://github.com/babel/babel/pull/13416))

["Hack-style pipelines"](https://github.com/js-choi/proposal-hack-pipes/) is a new flavor of the [pipeline operator](https://github.com/tc39/proposal-pipeline-operator) proposal, intended to replace the "smart mix" variant.

Hack-style pipelines require you to _always_ use a "topic token" (such as `#`) to reference the value of the previous pipeline step:

```javascript
// Input
"World"
  |> `Hello, ${#}!`
  |> alert(#);

// output
var _ref, _ref2;

_ref2 = (_ref = "World", `Hello, ${_ref}!`), alert(_ref2);
```

> You can test this proposal by enabling the [`version: "hack"`](TODO: Link docs) option in `@babel/plugin-proposal-pipeline-operator`. You must also choose which topic token to use, between `"#"` and `"%"`:

Author: ```suggestion
You can test this proposal by enabling the [`proposal: "hack"`](https://babeljs.io/docs/en/babel-plugin-proposal-pipeline-operator#usage) option in `@babel/plugin-proposal-pipeline-operator`. You must also choose which topic token to use, between `"#"` and `"%"`:
```

```json
// babel.config.json
{
  "plugins": [
    ["@babel/plugin-proposal-pipeline-operator", {

>       "version": "hack",

Author: ```suggestion
      "proposal": "hack",
```

---
layout: post
title:  "7.15.0 Released: Hack-style pipelines, TypeScript const enums and Rhino target support"
author: Babel Team
authorURL: https://twitter.com/babeljs
date:   2021-07-26 0:00:00
categories: announcements
share_text: "Babel 7.15.0 Released"
---

> This release enables top-level await parsing by default (Stage 4 at the [May](https://github.com/tc39/agendas/blob/master/2021/05.md) meeting), enables transforming ergonomic brand checks for private fields by default (Stage 4 at the [July](https://github.com/tc39/agendas/blob/master/2021/07.md) meeting) and includes support for the [Hack-style](https://github.com/js-choi/proposal-hack-pipes) pipeline operator. We also improved our TypeScript support, implementing transform support for `const` enums and namespace aliases, and we expanded our heuristics to add `.displayName` to React components created by `React.createContext()` ([#13501](https://github.com/babel/babel/pull/13501)).

Reviewer 3: ```suggestion
This release enables parsing top-level await (Stage 4 at the [May](https://github.com/tc39/agendas/blob/master/2021/05.md) meeting) and transforming ergonomic brand checks for private fields (Stage 4 at the [July](https://github.com/tc39/agendas/blob/master/2021/07.md) meeting) by default. There is also now support for the [Hack-style](https://github.com/js-choi/proposal-hack-pipes) pipeline operator. We also improved our TypeScript support, implementing transform support for `const` enums and namespace aliases, and we expanded our heuristics to add `.displayName` to React components created by `React.createContext()` ([#13501](https://github.com/babel/babel/pull/13501)).
```

We also introduced a new compiler assumption, [`noIncompleteNsImportDetection`](TODO: Link), to produce a smaller output when compiling ECMAScript modules to CommonJS without worrying about partially initialized namespace imports caused by module cycles.

Additionally, you can now specify [Rhino](https://mozilla.github.io/rhino/) as a compilation target.

You can read the whole changelog [on GitHub](https://github.com/babel/babel/releases/tag/v7.15.0).

<!-- truncate -->

> > Mention [recent blog post](https://babeljs.io/blog/2021/05/10/funding-update) and quick funding update?

Reviewer: ```suggestion
> Mention [recent blog post](https://babeljs.io/blog/2021/05/10/funding-update) and quick funding update?
We're really grateful for all the support the community has shown over the last month's since our [funding post](https://babeljs.io/blog/2021/05/10/funding-update) update in May.
```

Not sure what else to add, I guess some numbers like being able to sustain what we had with 6k/month for everyone, but we can do another update on all that later as well. Still looking for other kinds of funding opportunities, maybe we can mention participating in gitcoin grants or something?

## Highlights

### ECMAScript features enabled by default

In the last two meetings, the [top-level `await`](https://github.com/tc39/proposal-top-level-await) and [ergonomic brand checks for private fields](https://github.com/tc39/proposal-private-fields-in-in) proposals reached Stage 4.

```javascript
import * as db from "database";

await db.connect(); // top-level await

class DBConnector {
  #password;
  static isConnector(obj) {
    return #password in obj; // ergonomic brand checks
  }
}
```

Babel now supports them by default, so you can remove the following plugins from your configuration:
- `@babel/plugin-syntax-top-level-await`
- `@babel/plugin-syntax-private-property-in-object`
- `@babel/plugin-proposal-private-property-in-object`

> Please note that Babel cannot transform top-level `await`, but it can only parse it.

Reviewer 3: ```suggestion
Please note that Babel can currently only parse top-level `await` and cannot transform it.
```

### New TypeScript features ([#13324](https://github.com/babel/babel/pull/13324), [#13528](https://github.com/babel/babel/pull/13528))

[TypeScript 4.4](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/) doesn't include any new syntax that we had to implement, other than a minor restriction: you cannot specify the value of an abstract class field.

```typescript=
abstract class C {
    abstract prop = 1; // now a SyntaxError!
}
```

However, we did add two TypeScript features that we have been missing for a long time: `const` enums and namespace aliases (`import Alias = Namespace`).

> Previously we presented an error when using `const` enums since it requires type information to compile correctly. As a workaround, the community built plugins such as [`babel-plugin-const-enum`](https://github.com/dosentmatter/babel-plugin-const-enum#readme). So Babel now ignores the `const` modifier when compiling enums: this matches TypeScript's behavior when using the `--isolatedModules` options.

Reviewer 3: ```suggestion
Previously we presented an error when using `const` enums since it requires type information to compile correctly. As a workaround, the community built plugins such as [`babel-plugin-const-enum`](https://github.com/dosentmatter/babel-plugin-const-enum#readme). Babel now ignores the `const` modifier when compiling enums, which matches TypeScript's behavior when using the `--isolatedModules` option.
```

If you want a more optimized output similar to the default code produced by TypeScript, you can enable the [`optimizeConstEnums`](TODO: Link docs) option of `@babel/plugin-tranform-typescript` or `@babel/preset-typescript`.

```typescript
// Input
const enum Animals { Dog }
console.log(Animals.Dog);

// Output (default)
var Animals;
(function (Animals) {
  Animals[Animals["Dog"] = 0] = "Dog";
})(Animals || (Animals = {}));

console.log(Animals.Dog);

// Output with `optimizeConstEnums`
console.log(0);
```
### Hack-style pipeline operator support ([#13191](https://github.com/babel/babel/pull/13191), [#13416](https://github.com/babel/babel/pull/13416))

["Hack-style pipelines"](https://github.com/js-choi/proposal-hack-pipes/) is a new flavor of the [pipeline operator](https://github.com/tc39/proposal-pipeline-operator) proposal, intended to replace the "smart mix" variant.

> Hack-style pipelines require you to _always_ use a "topic token" (such as `#`), to reference the value of the previous pipeline step:

Reviewer 3: ```suggestion
Hack-style pipelines require you to _always_ use a "topic token" (such as `#`) to reference the value of the previous pipeline step:
```

```javascript
// Input
"World"
  |> `Hello, ${#}!`
  |> alert(#);

// output
var _ref, _ref2;

_ref2 = (_ref = "World", `Hello, ${_ref}!`), alert(_ref2);
```

You can test this proposal by enabling the [`version: "hack"`](TODO: Link docs) option in `@babel/plugin-proposal-pipeline-operator`. You must also choose which topic token to use, between `"#"` and `"%"`:

```json
// babel.config.json
{
  "plugins": [
    ["@babel/plugin-proposal-pipeline-operator", {
      "version": "hack",
      "topicToken": "#"
    }]
  ]
}
```

### Preparing `@babel/eslint-parser` for Babel 8 ([#13398](https://github.com/babel/babel/pull/13398))

We have been slowly continuing to work on Babel 8 in the past months. We are not ready for a Babel 8 beta release yet, but we are starting to expose the first experimental changes.

We plan to fully convert Babel from CommonJS to ECMAScript modules, but this has a problem: configuration loading will be asynchronous _more often_, and `@babel/eslint-parser` can only work synchronously (because ESLint only supports synchronous parsers).

> `@babel/eslint-parser` 7.15.0 exposes a new entry-point: `@babel/eslint-parser/experimental-worker`. It moves the Babel config loading and paring task to a separade worker which is managed synchronously from the main thread, while still supporting async configuraion loading. As an immediate advantage for Babel 7, it allows using native ECMAScript modules for Babel configuration files even when using `@babel/eslint-parser`.

Reviewer 4: typo:
```suggestion
`@babel/eslint-parser` 7.15.0 exposes a new entry-point: `@babel/eslint-parser/experimental-worker`. It moves the Babel config loading and paring task to a separate worker which is managed synchronously from the main thread, while still supporting async configuraion loading. As an immediate advantage for Babel 7, it allows using native ECMAScript modules for Babel configuration files even when using `@babel/eslint-parser`.
```

> `@babel/eslint-parser` 7.15.0 exposes a new entry-point: `@babel/eslint-parser/experimental-worker`. It moves the Babel config loading and paring task to a separate worker which is managed synchronously from the main thread, while still supporting async configuraion loading. As an immediate advantage for Babel 7, it allows using native ECMAScript modules for Babel configuration files even when using `@babel/eslint-parser`.

Reviewer 3: ```suggestion
`@babel/eslint-parser` 7.15.0 exposes a new entry-point: `@babel/eslint-parser/experimental-worker`. It moves the Babel config loading and parsing task to a separate worker which is managed synchronously from the main thread, while still supporting async configuration loading. As an immediate advantage for Babel 7, it allows using native ECMAScript modules for Babel configuration files even when using `@babel/eslint-parser`.
```
